File: Classwork/Lecture10/Lecture_10_Torch.r
library(torch)

if (cuda_is_available()) {
    device <- torch_device("cuda")
} else {
    device <- torch_device("cpu")
}

# splitting IRIS
idx <- sample(nrow(iris), nrow(iris) * 0.8)

x_train <- as.matrix(iris[idx, 1:4])
y_train <- as.integer(iris[idx, 5])

x_test <- as.matrix(iris[-idx, 1:4])
y_test <- as.integer(iris[-idx, 5])

# convert to tensors
x_train_tensor <- torch_tensor(x_train, dtype = torch_float())
y_train_tensor <- torch_tensor(y_train, dtype = torch_long())

x_test_tensor <- torch_tensor(x_test, dtype = torch_float())
y_test_tensor <- torch_tensor(y_test, dtype = torch_long())

model <- nn_sequential( #4/8/16/3
    # layer 1
    nn_linear(4, 8), nn_relu(),
    # layer 2
    nn_linear(8, 16), nn_relu(),
    # layer 3
    nn_linear(16, 3), nn_softmax(dim = 2)
)

# Define cost function and optimizer
criterion <- nn_cross_entropy_loss()
optimizer <- optim_adam(model$parameters, lr = 0.005)

epochs <- 300

for (i in 1:epochs) {
    optimizer$zero_grad()

    # Forward pass
    y_pred_tensor <- model(x_train_tensor)

    # Compute loss
    loss <- criterion(y_pred_tensor, y_train_tensor)
    loss$backward()

    # take a step in the opposite direction
    optimizer$step()

    if (i %% 10 == 0) {
        winners <- y_pred_tensor$argmax(dim = 2)
        corrects <- winners == y_train_tensor
        accuracy <- corrects$sum()$item() / y_train_tensor$size()
        cat("Epoch:", i,
            "Loss", loss$item(),
            "Accuracy", accuracy, "\n")
    }
}

# Check on the test set
y_pred_tensor <- model(x_test_tensor)
y_pred <- as.array(y_pred_tensor$argmax(dim = 2))

print(table(y_pred, y_test))
cat(" Accuracy: ", sum(y_pred == y_test) / length(y_pred), "\n")

# # How to create a net by using modules (classes) instead of nn_sequential

# net <- nn_module(
#   "class_net",

#   initialize = function() {
#     self$linear1 <- nn_linear(4, 8)
#     self$linear2 <- nn_linear(8, 16)
#     self$linear3 <- nn_linear(16, 3)
#   },

#   forward = function(x) {
#     x |>
#     self$linear1() |>
#     nnf_relu() |>
#     self$linear2() |>
#     nnf_relu() |>
#     self$linear3() |>
#     nnf_softmax(dim = 2)
#   }
# )

# # Instantiate an object of the class
# model <- net()

File: Classwork/Lecture10/Lecture_10_Torch.py
import numpy as np
import pandas as pd
import torch
import torch.nn as nn
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix

if torch.cuda.is_available():
    device = torch.device("cuda")
else:
    device = torch.device("cpu")

# Read dataset
iris = pd.read_csv('./Lecture_10_iris.csv')

# Label encoding
iris['Species'] = iris['Species'].astype('category').cat.codes

x_train, x_test, y_train, y_test = train_test_split(iris[iris.columns[0:4]].values,
                                                    iris.Species.values, test_size=0.2)

# convert to tensors
x_train_tensor = torch.Tensor(x_train).float()
y_train_tensor = torch.Tensor(y_train).long()

x_test_tensor = torch.Tensor(x_test).float()
y_test_tensor = torch.Tensor(y_test).long()


model = nn.Sequential(
    # layer 1
    nn.Linear(4, 8), nn.ReLU(),  # 4/8/16/3
    # layer 2
    nn.Linear(8, 16), nn.ReLU(),  # 4/8/16/3
    # layer 3
    nn.Linear(16, 3), nn.Softmax(dim = 1)  # 4/8/16/3
)

# Define cost function and optimizer
criterion = nn.CrossEntropyLoss()  # combines nn.LogSoftmax() and nn.NLLLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.005)  # lr = learning rate

epochs = 300  # number of times to loop through the entire dataset

for i in range(epochs):  # loop through the entire dataset a number of times
    optimizer.zero_grad()  # zero the gradients

    # Forward pass: Compute predicted y by passing  x to the model
    y_pred_tensor = model(x_train_tensor)  # model.forward(x_train_tensor)

    # Compute loss
    loss = criterion(y_pred_tensor, y_train_tensor)
    loss.backward()
    
    # take a step in the opposite direction
    optimizer.step()

    if i % 10 == 0:
        winners = y_pred_tensor.argmax(dim=1)
        print('Epoch:', i, 
              'Loss:', loss.item(), 
              'Accuracy:', accuracy_score(winners, y_train_tensor))


y_pred_tensor=model(x_test_tensor)
# print(y_pred_tensor)

y_pred = y_pred_tensor.argmax(1)
# _, y_pred = y_pred_tensor.max(1)

print('\nConfusion Matrix (Test Set):\n', confusion_matrix(y_pred, y_test))
print('\nAccuracy (Test Set):', accuracy_score(y_test_tensor, y_pred))

# class Net(nn.Module):
#   def __init__(self, n_input, n_layer1, n_layer2, n_output):
#     super().__init__()
#     self.hid1 = nn.Linear(n_input, n_layer1)  # 4/8/16/3
#     self.relu1 = nn.ReLU()
#     self.hid2 = nn.Linear(n_layer1, n_layer2)
#     self.relu2 = nn.ReLU()
#     self.oupt = nn.Linear(n_layer2, n_output)
#     self.softmax = nn.Softmax(dim=1)

#   def forward(self, x):
#     z = self.hid1(x)
#     z = self.relu1(z)
#     z = self.hid2(z)
#     z = self.relu2(z)
#     z = self.oupt(z)  
#     z = self.softmax(z)
#     return z

# model = Net(n_input = 4, n_layer1 = 8, n_layer2 = 16, n_output = 3)